### PR TITLE
[MC][x86] Allow non-MCTargetExpr RHS when the LHS of a MCBinaryExpr is MCTargetExpr

### LINKED ISSUE #73109: clang: 18: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
When compile this program with clang 18, clang 18 assertion fails, while clang 17 compiles it without any issues..
```
void f(unsigned char *a) {
  int ret;
  asm volatile("setb %0 %1" : "=r"(ret) : "r"(a));
}
```

Compiler Explorer: https://gcc.godbolt.org/z/G664rETEz

The crash output:
```
clang: /root/llvm-project/llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From*) [with To = llvm::MCTargetExpr; From = const llvm::MCExpr]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/clang -gdwarf-4 -g -o /app/output.s -S --gcc-toolchain=/opt/compiler-explorer/gcc-9.2.0 -fcolor-diagnostics -fno-crash-diagnostics -O0 <source>
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '<source>'.
4.	Running pass 'X86 Assembly Printer' on function '@f'
 #0 0x00000000037619f8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x37619f8)
 #1 0x000000000375f6dc llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x375f6dc)
 #2 0x00000000036a8248 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007fc2c93bb420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #4 0x00007fc2c8e7e00b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #5 0x00007fc2c8e5d859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #6 0x00007fc2c8e5d729 (/lib/x86_64-linux-gnu/libc.so.6+0x22729)
 #7 0x00007fc2c8e6efd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #8 0x00000000034548df llvm::MCExpr::evaluateAsRelocatableImpl(llvm::MCValue&, llvm::MCAssembler const*, llvm::MCAsmLayout const*, llvm::MCFixup const*, llvm::DenseMap<llvm::MCSection const*, unsigned long, llvm::DenseMapInfo<llvm::MCSection const*, void>, llvm::detail::DenseMapPair<llvm::MCSection const*, unsigned long>> const*, bool) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x34548df)
 #9 0x0000000003454b38 llvm::MCExpr::evaluateAsAbsolute(long&) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x3454b38)
#10 0x00000000034e8adb (anonymous namespace)::AsmParser::parseExpression(llvm::MCExpr const*&, llvm::SMLoc&) (.part.0) AsmParser.cpp:0:0
#11 0x00000000025febf6 (anonymous namespace)::X86AsmParser::parseATTOperand(llvm::SmallVectorImpl<std::unique_ptr<llvm::MCParsedAsmOperand, std::default_delete<llvm::MCParsedAsmOperand>>>&) X86AsmParser.cpp:0:0
#12 0x0000000002603ae1 (anonymous namespace)::X86AsmParser::ParseInstruction(llvm::ParseInstructionInfo&, llvm::StringRef, llvm::SMLoc, llvm::SmallVectorImpl<std::unique_ptr<llvm::MCParsedAsmOperand, std::default_delete<llvm::MCParsedAsmOperand>>>&) X86AsmParser.cpp:0:0
#13 0x00000000034e12d2 (anonymous namespace)::AsmParser::parseAndMatchAndEmitTargetInstruction((anonymous namespace)::ParseStatementInfo&, llvm::StringRef, llvm::AsmToken, llvm::SMLoc) AsmParser.cpp:0:0
#14 0x00000000034fa327 (anonymous namespace)::AsmParser::parseStatement((anonymous namespace)::ParseStatementInfo&, llvm::MCAsmParserSemaCallback*) (.part.0) AsmParser.cpp:0:0
#15 0x00000000034eeb85 (anonymous namespace)::AsmParser::Run(bool, bool) AsmParser.cpp:0:0
#16 0x00000000045dfcda llvm::AsmPrinter::emitInlineAsm(llvm::StringRef, llvm::MCSubtargetInfo const&, llvm::MCTargetOptions const&, llvm::MDNode const*, llvm::InlineAsm::AsmDialect) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x45dfcda)
#17 0x00000000045e0b99 llvm::AsmPrinter::emitInlineAsm(llvm::MachineInstr const*) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x45e0b99)
#18 0x00000000045dbed0 llvm::AsmPrinter::emitFunctionBody() (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x45dbed0)
#19 0x0000000002270661 llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x2270661)
#20 0x0000000002b5ffe1 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
#21 0x000000000311a409 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x311a409)
#22 0x000000000311a641 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x311a641)
#23 0x000000000311ae82 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x311ae82)
#24 0x00000000039d7046 (anonymous namespace)::EmitAssemblyHelper::EmitAssembly(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) BackendUtil.cpp:0:0
#25 0x00000000039d73d1 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x39d73d1)
#26 0x0000000003fe10c8 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x3fe10c8)
#27 0x0000000005efcf09 clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x5efcf09)
#28 0x0000000003fe05e8 clang::CodeGenAction::ExecuteAction() (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x3fe05e8)
#29 0x0000000004246f89 clang::FrontendAction::Execute() (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x4246f89)
#30 0x00000000041c74ae clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x41c74ae)
#31 0x000000000432653e clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x432653e)
#32 0x0000000000bfd846 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0xbfd846)
#33 0x0000000000bf510a ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#34 0x0000000004021909 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#35 0x00000000036a86f4 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x36a86f4)
#36 0x0000000004021eff clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#37 0x0000000003fe9e75 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x3fe9e75)
#38 0x0000000003fea8dd clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x3fea8dd)
#39 0x0000000003ff2805 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0x3ff2805)
#40 0x0000000000bfacec clang_main(int, char**, llvm::ToolContext const&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0xbfacec)
#41 0x0000000000af4381 main (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0xaf4381)
#42 0x00007fc2c8e5f083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#43 0x0000000000bf4bee _start (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+0xbf4bee)
clang: error: clang frontend command failed with exit code 134 (use -v to see invocation)
Compiler returned: 134
```

### PR DESCRIPTION
**Pull request by cygao90**
This fixes #73109.
In instruction `addl %eax %rax`, because there is a missing comma in the middle of two registers, the asm parser will treat it as a binary expression.
```
%rax  %  rax --> register mod identifier
```
However, In `MCExpr::evaluateAsRelocatableImpl`, it only checks the left side of the expression. This patch ensures the right side will also be checked.

### CODE REVIEW

**Comment by phoebewang** [comment]
Thanks for the fix! I think it would be a typo when added the code in https://github.com/llvm/llvm-project/commit/7fd992a75502b7eb251daa1b1e4d23a73fa76d53
But I'd like to invite @rnk to double confirm it.

**Review by phoebewang — CHANGES_REQUESTED** [review]
I found a problem when built it locally:

```
$ echo 'setb %al %rbx' | llvm-mc --show-encoding
        .text
        setb    (%al)%rbx                       # encoding: [0x0f,0x92,0x04,0x25,A,A,A,A]
                                        #   fixup A - offset: 4, value: (%al)%rbx, kind: reloc_signed_4byte
```

In this case, `%rbx` is taken as a relocation, which is incorrect.
I think the `cast` is intended here. What we should do is to make it user friendly by emitting an error with message. But it seems not easy to do here.

**Comment by cygao90** [comment]
@phoebewang Thank you for your review. I tested on gcc. gcc will also accept such instruction:
https://godbolt.org/z/er5zEhThY
but assembler will reject it:
https://godbolt.org/z/4EnPq7rjq
Perhaps we should leave it to llvm assembler
```
echo 'setb %al %rbx' | ./llvm-mc --filetype=obj -o test
<stdin>:1:1: error: expected relocatable expression
setb %al %rbx
^
```

**Comment by MaskRay** [comment]
I fixed `MCBinaryExpr::NE` and extended `llvm/test/MC/X86/register-assignment.s`, but did not think much about non-MCTargetExpr `RHS`.

I think this patch is correct but we need a test to `llvm/test/MC/X86/register-assignment.s`. It can be `.if var_cata == 1 ...`

**Review by phoebewang — APPROVED** [review]
Makes sense to me, please address @MaskRay 's comments.

**Comment by cygao90** [comment]
I don't know why this check failed...

**Comment by phoebewang** [comment]
Check fail occurs sometimes for other reasons like infrastructure issue or problem in recent merged patches. You just need to take a look at the failed tests and ignore it if they are unrelated to your change.

**Comment by cygao90** [comment]
```
--
  | MLIR :: python/dialects/affine.py
  | MLIR :: python/dialects/amdgpu.py
  | MLIR :: python/dialects/arith_dialect.py
  | MLIR :: python/dialects/builtin.py
  | MLIR :: python/dialects/cf.py
  | MLIR :: python/dialects/complex_dialect.py
  | MLIR :: python/dialects/func.py
  | MLIR :: python/dialects/linalg/opdsl/arguments.py
  | MLIR :: python/dialects/linalg/opdsl/assignments.py
  | MLIR :: python/dialects/linalg/opdsl/doctests.py
  | MLIR :: python/dialects/linalg/opdsl/emit_convolution.py
  | MLIR :: python/dialects/linalg/opdsl/emit_fill.py
  | MLIR :: python/dialects/linalg/opdsl/emit_matmul.py
  | MLIR :: python/dialects/linalg/opdsl/emit_misc.py
  | MLIR :: python/dialects/linalg/opdsl/emit_pooling.py
  | MLIR :: python/dialects/linalg/opdsl/metadata.py
  | MLIR :: python/dialects/linalg/opdsl/shape_maps_iteration.py
  | MLIR :: python/dialects/linalg/opdsl/test_core_named_ops.py
  | MLIR :: python/dialects/linalg/ops.py
  | MLIR :: python/dialects/math_dialect.py
  | MLIR :: python/dialects/memref.py
  | MLIR :: python/dialects/ml_program.py
  | MLIR :: python/dialects/nvgpu.py
  | MLIR :: python/dialects/nvvm.py
  | MLIR :: python/dialects/ods_helpers.py
  | MLIR :: python/dialects/pdl_ops.py
  | MLIR :: python/dialects/python_test.py
  | MLIR :: python/dialects/quant.py
  | MLIR :: python/dialects/scf.py
  | MLIR :: python/dialects/shape.py
  | MLIR :: python/dialects/tensor.py
  | MLIR :: python/dialects/transform.py
  | MLIR :: python/dialects/transform_bufferization_ext.py
  | MLIR :: python/dialects/transform_structured_ext.py
  | MLIR :: python/dialects/vector.py
  | MLIR :: python/ir/array_attributes.py
  | MLIR :: python/ir/attributes.py
  | MLIR :: python/ir/blocks.py
  | MLIR :: python/ir/builtin_types.py
  | MLIR :: python/ir/diagnostic_handler.py
  | MLIR :: python/ir/dialects.py
  | MLIR :: python/ir/location.py
  | MLIR :: python/ir/operation.py
  | MLIR :: python/ir/symbol_table.py
  | MLIR :: python/ir/value.py
  | MLIR :: python/pass_manager.py
```
Seems unrelated.

File: llvm/test/MC/X86/register-assignment-error.s
Comment (KanRobert) [inline llvm/test/MC/X86/register-assignment-error.s:1]: Minor suggestion

You don't need `-unknown-unknown`, `x86_64` is enough. Ony one check prefix is used, `--check-prefix=CHECK-ERROR` is redundant.

**Review by KanRobert — APPROVED** [review]

**Comment by cygao90** [comment]
Can someone with write access merge this PR if no further change is required? Thanks.

**Comment by MaskRay** [comment]
> Can someone with write access merge this PR if no further change is required? Thanks.

We should add a test `.ifdef 1 == var_xdata` as well and ensure there is an EOL at the end of the file. I merged the new test file into `register-assignment.s` in a22c8efad733cf3407a93387a6a9c743659dc74c.

**Pull request merged**
